content/blogposts/2024/anatomy-of-a-coding-assistant.md

```diff
@@ -0,0 +1,157 @@
+authors:
+  - name: YK Sugi
+    url: https://x.com/ykdojo
+publishDate: 2024-06-13T10:00-01:00
+description: "Discover how AI coding assistants like Sourcegraph Cody enhance developer productivity by leveraging context. Learn about features like autocomplete, chat, test generation, and code editing, and how they fetch relevant code snippets to provide accurate suggestions and edits."
+tags: [blog]
+slug: 'content/blogposts/2024/anatomy-of-a-coding-assistant'
+published: true
+heroImage: https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/1_pointer.png
+socialImage: https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/1_pointer.png
+---
+
+![1_pointer](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/1_pointer.png)
+
+Have you ever wondered how an AI coding assistant works?
+
+We’ve been working on our AI coding assistant, Cody, for a while now, and over time we’ve learned that really, ***context is king***.
+
+![2_octopus](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/2_octopus.png)
+
+If you ask ChatGPT a coding question about your particular codebase, it won’t be able to answer it because it doesn’t know anything about your codebase.
+
+![3_chatgpt](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/3_chatgpt.png)
+
+Even if it happens to be open source, the answer you get might be outdated or not accurate enough.
+
+![4_chatgpt2](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/4_chatgpt2.png)
+
+So, by fetching the right parts of your codebase and giving them to your AI coding assistant, the underlying AI model is able to come up with much better answers.
+
+![5_cody](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/5_cody.png)
+
+Over the past year or so, we’ve learned a lot about what works for fetching the right context for an AI coding assistant. In this post, I’m going to share some of those learnings and uncover how it works under the hood.
+
+---
+
+## An AI coding assistant - a product of products
+
+The reality is, if you look at a coding assistant like [Sourcegraph Cody](https://sourcegraph.com/cody), it’s not a single product—it’s many products put together. You have autocomplete, which makes suggestions as you code that keep you in flow. Then there’s chat, which allows you to ask questions of and generate code that fits in your codebase. Cody also has a feature that automates quick inline code changes with natural language, and another one that generates unit tests.
+
+![6_edits](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/6_edits.png)
+
+Each of these features works very differently from one another. The mental model of the user and the requirements are different for each. For autocomplete, speed is key. You want it to be very fast and accurate enough so that it doesn’t interrupt the programmer’s flow. On the other hand, for chat, accuracy and the quality of the responses may be more important.
```

Comment:
```suggestion
Each of these features works very differently from one another. The mental model of the user and the requirements are different for each. **For autocomplete, speed is key.** You want it to be very fast and accurate enough so that it doesn’t interrupt the programmer’s flow. On the other hand, **for chat, accuracy and the quality of the responses may be more important**.
```